Cover Art Fetcher

fatihemreyildiz opened the pull request:
Hey everyone!

My goal is to get cover arts of the tracks by using `coverartarchive` on the "Import Metadata from Musicbrainz" menu, aka Tag Fetcher Menu. I am working on this feature right now. This is the initial PR to get your ideas-thoughts and of course feedback about the GUI of this feature. More information about this PR can be found on our Zulip chat https://mixxx.zulipchat.com/#narrow/stream/109215-gsoc/topic/cover.20art.20lookup

This is the first interaction for now. I have displayed the existing cover art of the track on the Tag Fetcher Menu. The Tag Fetcher Menu seems like this. ![Tag Fetcher Menu](https://user-images.githubusercontent.com/67206006/178327695-4e9e0fb1-dc8c-4fc0-bc48-2672028f86fd.png)

Right now I can send the metadata of the Suggested track to coverartarchive and get a response via a button. It is better to first decide what should GUI look like? Then we can implement that (or something similar) later on. 

For to decide about GUI, we should decide how this feature can be used?
* I think there should be a button called "look for cover art" which triggers the request when pressed, and can get cover art of the selected `Suggested Tags`.
* Should we have separate buttons for the metadata and cover art? Can cover-art be updated separate then the metadata or apply button should update both the cover art and the metadata? 

Do you have any other use-case comes to your mind? These few buttons would make this menu crowded and complicated? What do you think generally? 

(This feature also needs your feedback about technical stuff, the first thing is how to store the cover-arts, cover-arts has thumbnails should we use them, if the cover art updated, should we save it as a file or embed in the file, should user choose between these storing options and so on...)

fatihemreyildiz committed:
coverartfetcher: update the latest status.

ronso0 commented:
Nice!

The GUI can be as rough as necessary for testing the functionality, don't worry. That can be rearranged and polished later on.

Just some ideas/impressions regarding the GUI:
* show the current cover and new candidates in a vertical layout:
  * "current:" at the top
  * candidate (blank cover if no candidate has been fetched?)
  * "Get cover art" button at the bottom?
* maybe it would be worth to rework the WCoverArt instantiation so the right-click menu can be disabled in the musicbrainz dialog? 

I think we should fetch an adequate thumbnail first. Smallest is 250px, right? Then try choose bigger version if necessary, depending on the Mixxx GUI scalefactor, so images look sharp on HiDPI screens.
Maybe users want to preview the image in full size. Usually left-click on the cover opens the separate resizable popup. Dunno how that would feel like if the full-size cover needs to be fetched first.

fatihemreyildiz commented:
Thanks you for the feedback @ronso0 

> The GUI can be as rough as necessary for testing the functionality, don't worry. That can be rearranged and polished later on.

That's good! So we must decide the GUI at least for the testing. 

>show the current cover and new candidates in a vertical layout:
"current:" at the top
candidate (blank cover if no candidate has been fetched?)
"Get cover art" button at the bottom?

I think that 2 cover arts is a great idea. Both for technical and UX way. So these two cover arts can have different functionality. One is simply can show the existing cover art, the other can show selected `suggested tags`.

> maybe it would be worth to rework the WCoverArt instantiation so the right-click menu can be disabled in the musicbrainz dialog?

Maybe a child class can do that. Which can only show the cover art and populate the full size of the cover art. Would something like that work?

>I think we should fetch an adequate thumbnail first. Smallest is 250px, right? Then try choose bigger version if necessary, depending on the Mixxx GUI scalefactor, so images look sharp on HiDPI screens.
Maybe users want to preview the image in full size. Usually left-click on the cover opens the separate resizable popup. Dunno how that would feel like if the full-size cover needs to be fetched first.

Yeah thumbnails are 250px. I think for displaying the `suggested tag` cover arts we better use the 250px in order to reduce the performance issues. We thought users can choose the resolution of the cover art. We have a popup for existing cover-arts, maybe something similar can be done for the founded cover arts.

By the way, I was playing with the UI, so I did something like this. With this one button, maybe found cover arts can be displayed if the selected result triggered. What do you think?
![CoverArtFetcher](https://user-images.githubusercontent.com/67206006/178361343-66166a40-7d7c-467e-9924-c09dada794c4.png)

daschuer commented:
The resolution/quality question has some aspects we need to consider.
* Internet bandwidth
* Display resolution 
* Storage space 
* Time for fetching 
* rate limits 

What is the best strategy? 

We may either always fetch all thumbnail covers for all MusicBrainz results, so that they are immediately displayed when selecting a row. When the user has decided to integrate a cover, we may download the desired target size. 

Another strategy may be to download only the desired target size on demand and rescale it before displaying. 

What should be the default behaviour? What should be the default size? 

@ronso0 what do you think?

daschuer commented:
I can think of a few checkboxes:
* [ ] Fetch Cover 
* [ ] Use High-Res cover (1MB)
* [ ] Store as track-file-name.jpg
* [ ] Integrate in track file 

Is this a good Idea to offer these options?
Do we want to place them on the dialog or in preference?

Swiftb0y commented:
Does Qt have some abstraction that we can query whether the system is on a metered connection? We could use that for deciding the quality option and eagerness of the fetching. So we can save bandwidth when the connection is metered.

fatihemreyildiz commented:
> Does Qt have some abstraction that we can query whether the system is on a metered connection? We could use that for deciding the quality option and eagerness of the fetching. So we can save bandwidth when the connection is metered.

I found [this](https://doc-snapshots.qt.io/qt6-dev/qnetworkinformation.html#isMetered-prop) on QT Doc, can it work? I assume it is not because It says since 6.2.

fatihemreyildiz commented:
> Thats pretty much exactly what I wanted, but yes, we can't use it because we're still stuck on Qt5.15

So there should be some TO DO comment lines for the future.

fatihemreyildiz commented:
> The resolution/quality question has some aspects we need to consider.
> 
> * Internet bandwidth
> * Display resolution
> * Storage space
> * Time for fetching
> * rate limits
> 
> What is the best strategy?
> 
> We may either always fetch all thumbnail covers for all MusicBrainz results, so that they are immediately displayed when selecting a row. When the user has decided to integrate a cover, we may download the desired target size.
> 
> Another strategy may be to download only the desired target size on demand and rescale it before displaying.
> 
> What should be the default behaviour? What should be the default size?
> 
> @ronso0 what do you think?

There is no rate limits according to Coverartarchive API documentation. 

That was exactly what I thought in the first place, we should fetch all thumbnails with the MusicBrainz results, and user can see by just clicking on the suggested tags, if they like they can get the cover art. 

I have tried the API with the few songs, most of the results are not stable. What I mean by that is there is a main image with the high resolution which was 11MB. Then thumbnails come along, the sizes are generally starts from 1200 and then gets smaller to 500 and 250. There are also 2 keys with the `small` and `large` which is equivalent of 250px and 500px. Some Album Release ID's have only small and large. So maybe we might use to fetch `small` for the first time and users can choose `large` to integrate. 

What do you think?

>I can think of a few checkboxes:
 Fetch Cover
 Use High-Res cover (1MB)
 Store as track-file-name.jpg
 Integrate in track file
Is this a good Idea to offer these options?
Do we want to place them on the dialog or in preference?

I think we should offer this options later on after this feature is working. And if we do that we should do in the preference. On preferences there could be section for cover art fetcher. If the users want to fetch cover arts they can choose. I think this will affect the UI, I couldn't make sure how to deal with it at the first time but this can be added as a first thing. Then resolution could be chosen on that screen to, small or large. Storage can be added in that section too.

Swiftb0y commented:
> So maybe we might use to fetch small for the first time and users can choose large to integrate.

Yes, use the most appropriate size for the UI (depending how big the coverart will be displayed) but use the best quality available for actually storing in mixxx.

daschuer commented:
11 MB for a cover is probably overdone for a 5 MB mp3 file. I cannot effort the HDD to double up the size of my collection because of cover arts. So is there a size that fits to the screen size when double clicking the cover? I think I can effort 200 kB or such for a cover. 
@Swiftb0y: Do we also need the option to actually use the 11 MB cover art? 
If we store it in a separate folder, the user can delete the big once if he is running out of HDD. Would that be worth to consider?

Swiftb0y commented:
Cover arts are usually only that big when their resolution is large (I'd estimate 11Mb to be about 5000x5000 pixels) or there is almost no compression... Rescaling/compressing the cover in mixxx is probably too complicated. I guess we can add a preference option for the upper limit in terms of size/resolution?

fatihemreyildiz commented:
I have checked with the few cover art responses. The original Image had big resolution, and it can change album release id to album release id. Small sizes (250px) are approx 20-30 KB and large size (500px) are 80-100 KB. 1200 px are approx 500kb. 

I think we should use small and large sizes for fetching and applying. I think large size enough for most use cases.

For the higher resolutions (1200px in this case), I'm not sure what to do. Also about the 11 Mb, I think that was only for an album, this can be changed and maybe could be more or less, so it is better not to use original image returned by the coverartarchive.

ronso0 commented:
> I think we should offer this options later on after this feature is working. And if we do that we should do in the preference. On preferences there could be section for cover art fetcher.

Yup. Btw I think we should try to split the Library page.

Swiftb0y commented:
We could also make the size of the cover art to be depended on the filesize. Then HQ flacs get HQ cover arts but LQ mp3s get low res cover arts. I think someone with a flac collection cares more about quality in general. Also since the size increase is relative, it wouldn't be that noticeable.

ronso0 commented:
> If we store it in a separate folder, the user can delete the big once if he is running out of HDD. Would that be worth to consider?

I wouldn't make it that complicated. A pref option [small | medium | large] as well as "Store in folder or add to file" should suffice.
For "medium" something between 800 and 1200 is okay, no? choose depending on what's available (if that info is available prior to fetching the image).

ronso0 commented:
> Then HQ flacs get HQ cover arts but LQ mp3s get low res cover arts. I think someone with a flac collection cares more about quality in general.

I do care about audio quality but I don't need 4k cover art : ) Thus I'd prefer a pref option.

ronso0 commented:
> I haven't understand what do you mean by split the library page, sorry. Could you please explain it more?

Just a side note, actually irrelevant here. I mean splitting up Preferences > Library into at least two sub-pages, because we added more options, thus the Library page gets longer and longer and needs to be scrolled and it becomes harder to find certain options.

fatihemreyildiz commented:
So, I have updated the GUI as I mentioned on [this comment](https://github.com/mixxxdj/mixxx/pull/4851#issuecomment-1180887696). I have left few comments in the new files/classes just for this commit to tell what I wanted to.

I think that on the Tag Fetcher menu it shouldn't be right clicked (cover art menu shouldn't be poped up). In order to do that, I needed another class. But the classes were almost the same, so in order to not duplicate the code, I have made a base class called "CoverArtLabel" (it is on the tag fetcher) and the widget one is derived from the base class. On my test it was working fine. But I'm not sure if I did something good. 

I might also need that for dlgcoverartfullsize, because it can also trigger the cover art menu, but in the future this menu can be made specialized to cover art fetcher such as, set this cover art, download this cover art and etc.

Also about the GUI, since I am not an expert, when I resize it, it breaks.

I think that these three parts need a bit refactor before moving to the technical part of this PR. 

Thanks in advance for the comments and feedback.

daschuer commented on src/library/coverartlabel.cpp:
```diff
@@ -0,0 +1,88 @@
+
+} // anonymous namespace
+
+CoverArtLabel::CoverArtLabel(QWidget* parent)
```
Since this is still a widget, I had expect to find this is the widget folder. 
We two names, one for the base without edit menu and one for the parent class with edit menu. 

How about 
WCoverArtLabelBase and WCoverArtLabelWithEditMenu

Or 

WCoverArtDisplay and WCoverArtDisplayEditable 


or such

@ronso0 what do yo think

fatihemreyildiz replied:
Only the menu is the difference, but I can not be sure about the future steps. Maybe later, just for "Fetched Cover Art" there could be a menu such as "apply this cover" or "find the album" etc.

> Though, as I said this is not high prio IMO and can be adressed later on.

:+1:

daschuer commented on src/library/dlgtagfetcher.ui:
```diff
@@ -80,6 +80,69 @@
+      <item row="4" column="1">
+      <widget class="QPushButton" name="btnCover">
+       <property name="text">
+        <string>&amp;Set Cover Art</string>
```
Is there a use case to use a cover form a different release than the actual release of the metadata? 
If yes, that should be probably rare. So I think I would conciser to just go with a single "Apply" button that applies all metadata at once.

fatihemreyildiz commented:
Since #4864 is merged, I have updated this PR.

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -222,14 +315,67 @@ void DlgTagFetcher::apply() {
 
+void DlgTagFetcher::retry() {
+    results->clear();
+    disconnect(m_track.get(),
```
Why is disconnecting and reconnect required?

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -222,14 +315,67 @@ void DlgTagFetcher::apply() {
+            this,
+            &DlgTagFetcher::slotTrackChanged);
+
+    addDivider(tr("Original tags"), results);
```
Can we get around this?

fatihemreyildiz replied:
If it is okay, I will mark these two conversation as resolved since we continue them on #4871 .

daschuer replied:
Yes, dies it make sence to rebase this on #4871?

fatihemreyildiz replied:
Yeah, that would be nice. should I wait for it to be in final situation?

daschuer commented on src/network/networktask.cpp:
```diff
@@ -66,6 +66,25 @@ void NetworkTask::emitAborted(
     emit aborted(requestUrl);
 }
 
+void NetworkTask::emitNotFound(QNetworkReply::NetworkError errorCode,
```
Can we avoid this extra signal? Is Webtask::onNetworkError() sufficient?

fatihemreyildiz replied:
As far as I remember, onNetworkError was related with the TagFetcher, so when this signal is emitted, Tag Fetcher Dialog refreshes itself, so the results from MusicBrainz were lost. I am not sure if this can be handled from slot of onNetworkError.

fatihemreyildiz commented:
Now this PR is rebased on #4871

fatihemreyildiz committed:
dlgtagfetcher.ui: fix wrong display.

fatihemreyildiz committed:
dlgtagfetcher: comment add to fetching user choice

fatihemreyildiz committed:
imagefiledata: improve format not matching warning

fatihemreyildiz committed:
dlgtagfetcher: polished & deleted redundants.

fatihemreyildiz committed:
dlgtagfetcher: CA constants added for progressbar

fatihemreyildiz committed:
cacworker: worker refactored for cover art fetcher

fatihemreyildiz committed:
dlgtagfetcher: cacopyworker implemented.

fatihemreyildiz commented:
I have implemented the worker into the `DlgTagFetcher`. 

It works fine according to my tests. As a use case, I didn't see any unwanted behaviour. But still, It needs a good refactoring. 

I think there are still these things refactored-added:
* Full size population for fetched cover art not populating properly. With my setup, I only have problem with Highest Size, If the fetched cover art more bigger than 1200x1200, it is not populated correctly.
* More debug messages can be added just in case of any failure in cover art tasks.
* Code can be refactored a bit, I added comments to make it explanatory and tried better naming, but a new perspective would be very helpful.

What do you think?

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+    QUuid selectedTagAlbumId = trackRelease.albumReleaseId;
+    statusMessage->setVisible(false);
+
+    QString coverArtMessage = tr("Looking for cover art");
+    loadingProgressBar->setVisible(true);
+    loadingProgressBar->setFormat(coverArtMessage);
+    loadingProgressBar->setValue(kPercentForCoverArtLinksTask);
```
```suggestion
    loadingProgressBar->setFormat(tr("Looking for cover art"));
    loadingProgressBar->setValue(kPercentForCoverArtLinksTask);
    loadingProgressBar->setVisible(true);
```

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+    }
+}
+
+void DlgTagFetcher::slotUpdateStatusMessage(const QString& message) {
```
unused?

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+
+    DlgPrefLibrary::CoverArtFetcherQuality fetcherQuality =
+            static_cast<DlgPrefLibrary::CoverArtFetcherQuality>(
+                    fetchedCoverArtQualityConfigValue);
```
This can be joint with the above.

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+        getCoverArt(allUrls.last());
+        return;
+    } else if (fetcherQuality == DlgPrefLibrary::CoverArtFetcherQuality::High) {
+        allUrls.size() < 3 ? getCoverArt(allUrls.last()) : getCoverArt(allUrls.at(2));
```
A check for 2 is missing

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+        allUrls.size() < 3 ? getCoverArt(allUrls.last()) : getCoverArt(allUrls.at(2));
+        return;
+    } else if (fetcherQuality == DlgPrefLibrary::CoverArtFetcherQuality::Medium) {
+        getCoverArt(allUrls.at(1));
```
A check for 1 is missing

fatihemreyildiz replied:
Normally all the responses at least have 2 cover art's URL, I haven't seen any response with one cover art URL.

So for all of the checks above would it be better to check if the allUrls is not empty and then the if block would do the rest? Will it work? What do you think?

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+        getCoverArt(allUrls.at(1));
+        return;
+    } else {
+        getCoverArt(allUrls.first());
```
an isEmpty() check is missing

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+    // Low     -> Always 250 PX fetched
+
+    if (fetcherQuality == DlgPrefLibrary::CoverArtFetcherQuality::Highest) {
+        getCoverArt(allUrls.last());
```
isEmpty() check is missing

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+    coverInfo.setImage();
+
+    loadingProgressBar->setVisible(false);
+    QString coverArtMessage = tr("Cover art is ready to be applied");
+    statusMessage->setVisible(true);
+    statusMessage->setText(coverArtMessage);
```
```suggestion
    statusMessage->setText(tr("Cover art is ready to be applied"));
    statusMessage->setVisible(true);
```

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+}
+
+void DlgTagFetcher::getCoverArt(const QString& url) {
+    QString coverArtMessage = tr("Cover art found getting image");
```
```suggestion
    QString coverArtMessage = tr("Cover art found, receiving image.");
```

daschuer commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -402,4 +521,162 @@ void DlgTagFetcher::tagSelected() {
+void DlgTagFetcher::slotCoverArtLinkNotFound() {
+    loadingProgressBar->setVisible(false);
+    statusMessage->setVisible(true);
+    QString message = tr("Cover Art is not available for selected tag");
```
```suggestion
    QString message = tr("Cover Art is not available for selected row");
```
or metadata, or release

daschuer commented on src/preferences/dialog/dlgpreflibrary.h:
```diff
@@ -20,6 +20,13 @@ class DlgPrefLibrary : public DlgPreferencePage, public Ui::DlgPrefLibraryDlg  {
         Ignore = 3,
     };
 
+    enum class CoverArtFetcherQuality : int {
```
```suggestion
    enum class CoverArtFetcherQuality {
```

daschuer reviewed (changes requested):
Here some comments, I need to test this as well. 
Sorry for the delay.

fatihemreyildiz committed:
dlgcoverartfullsize: special init renamed & tr add

fatihemreyildiz commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -82,6 +82,23 @@ void DlgCoverArtFullSize::init(TrackPointer pTrack) {
+
+    setWindowTitle(fetchedCoverArtWindowTitle);
+
+    //TODO: Do a better scaling. It populates wrong if the cover art size highest.
```
This issue only exists in the new code. Whenever the highest cover art is applied (now the cover art is not fetched but existed), nothing wrong with the scaling.

Such as: 
<img width="150" alt="FetchedCoverArt" src="https://user-images.githubusercontent.com/67206006/201518595-f0c03cd2-53f4-4abf-9a6e-8860fd80ca4c.png"> <img width="150" alt="AfterApplied" src="https://user-images.githubusercontent.com/67206006/201518597-3a8c66a3-a2db-4672-80a3-08207ed295b0.png">

fatihemreyildiz replied:
Both of the screen shot belongs the one fetched cover art but in different states, first screen shot when pressed on the fetched cover art (if the highest option is selected), it is not populated correctly, second one after it is applied (for the highest option), it populates just fine.

ronso0 replied:
Loading the data to `m_pixmap` solves the zoom issues I described.
For the actual resizing (adjusting to screen size) we need a lot of stuff from `DlgCoverArtFullSize::slotCoverFound` below, though this is not a stored cover, obviously we can't call that with all required arguments. Code duplication is pointless so maybe split out the resizing into a new function?

fatihemreyildiz committed:
dlgtagfetcher: polishing & check add for urls

fatihemreyildiz committed:
dlgtagfetcher: ui methods move to related .ui file

fatihemreyildiz commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -96,27 +114,41 @@ void updateOriginalTag(const Track& track, QTreeWidget* pParent) {
+    fetchedCoverArtLayout->setAlignment(Qt::AlignRight | Qt::AlignBottom | Qt::AlignCenter);
+    fetchedCoverArtLayout->setSpacing(0);
+    fetchedCoverArtLayout->setContentsMargins(0, 0, 0, 0);
+    fetchedCoverArtLayout->insertWidget(0, m_pWFetchedCoverArtLabel.get());
```
Only the insertWidget left now, the rest in the .ui file.

fatihemreyildiz committed:
tagfetcher: updated to latest musicbrainz fixes.

fatihemreyildiz commented:
Sorry for the delay, I tried to fix all the requested changes and also the conflicted files. 

As far as I tested, it works just fine, but still it needs still a good polishing, and tested in any way. 

What do you think? @daschuer @ronso0

daschuer commented:
Thank you for continuing her just after the prerequisites have landed in main. 

U have tested this a bit, and it is really a great feature. 

The jusability can be improved a bit. I am just not sure if we should hold this PR back over it. @ronso0 what do you think? 

Here my findings:
* The cover size (pixel count) varies for differnt covers. I like to add covers roughly at the same size. Was this just a bad pick or is the cover size unpredictable? Can we give an estimation what the sizes mean in terms of pixel and bytes? 
* The cover size setting is hard to discover. Can we make them accessible from the tag fetcher dialog? 
* I think we need a different empty cover, for not available vs. not set. 
* caching would be nice, for not fetching the same cover twice when scrolling though the results.

ronso0 commented:
I took a look and have these findings from a quick test with two tracks.
1. caching would indeed be good (at least for the time the dialog is open)
Is QPixmapCache sufficient?
2. no matter which image size I select in the preferences, clicking the cover preview has always the same size?
If I select "high" or "highest" I expect a 1:1 preview so I can judge the image quality
4. the coverart zoom is broken: the window resizes but the actual images doesn't (it's just cropped or padded with the background color, and the aspect ratio changes when zooming in - out)

ronso0 committed:
DlgCoverArtFullSize: adjust for population via CoverArtFetcher

ronso0 commented:
> the coverart zoom is broken: the window resizes but the actual images doesn't (it's just cropped or padded with the background color, and the aspect ratio changes when zooming in - out)

Fixed by 34c92ffa5239099649617049c9f6afcf4b529da1

ronso0 committed:
DlgPrefLibrary::slotResetToDefaults: remove obsolete actions

ronso0 commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -206,6 +206,10 @@ void DlgPrefLibrary::slotResetToDefaults() {
     radioButton_dbclick_bottom->setChecked(false);
     radioButton_dbclick_top->setChecked(false);
     radioButton_dbclick_deck->setChecked(true);
+    radioButton_cover_art_fetcher_highest->setChecked(false);
+    radioButton_cover_art_fetcher_high->setChecked(false);
+    radioButton_cover_art_fetcher_medium->setChecked(false);
+    radioButton_cover_art_fetcher_lowest->setChecked(true);
```
Since these radio buttons are in the same widget "groupBox_cover_art_fetcher" you only need to check one to disable the others. In slotUpdate() it's done correctly.
This also applies to the other radio buttons (Double-click action) where you copied this behavior from.

> Radio buttons are autoExclusive by default. If auto-exclusive is enabled, radio buttons that belong to the same parent widget behave as if they were part of the same exclusive button group. 
https://doc.qt.io/qt-6/qradiobutton.html#details

Feel free to pick 3f66b569a3

ronso0 commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -452,6 +452,66 @@
+   <item>
+    <widget class="QGroupBox" name="groupBox_cover_art_fetcher">
+     <property name="title">
+      <string>Cover Art Fetcher</string>
```
"Tag Fetcher: Preferred Cover Art Resolution"?
or
"Tag Fetcher: Preferred Cover Art Size"?
the the radio buttons could be labeled `XXL` / `XL` / `M` / `S` (is this more clear in terms of internationaization?) and would fit into one row.
Just an idea..

ronso0 replied:
dismissed, see my comment https://github.com/mixxxdj/mixxx/pull/4851#issuecomment-1321244868

ronso0 commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -452,6 +452,66 @@
+      <item row="4" column="0">
+       <widget class="QLabel" name="label_12">
+        <property name="text">
+         <string>Fetch cover art from coverartarchive.com by using Import Metadata From Musicbrainz.</string>
+        </property>
+       </widget>
+      </item>
+      <item row="5" column="0">
+       <widget class="QLabel" name="label_13">
+        <property name="text">
+         <string>Highest resolution can fetch up to very large cover arts. </string>
```
Those labels should go above the radio buttons IMO

ronso0 replied:
"Note: 'Highest' can fetch very large cover arts.

ronso0 commented:
Besides my findings above, this seems to work well.
Must have been a lot of work. Thanks @fatihemreyildiz for your endurance :+1:

ronso0 commented:
> no matter which image size I select in the preferences, clicking the cover preview has always the same size?
If I select "high" or "highest" I expect a 1:1 preview so I can judge the image quality

Okay, with some debug output and finally a track for which there is real HQ cover art available I can confirm **it works** as expected.

ronso0 commented:
> The cover size (pixel count) varies for differnt covers. I like to add covers roughly at the same size. Was this just a bad pick or is the cover size unpredictable? Can we give an estimation what the sizes mean in terms of pixel and bytes?

The  `json` result from coverartarchive (for example https://ia800805.us.archive.org/17/items/mbid-cf93b2c3-9152-489d-b879-772ca46f5d81/index.json) always contains at least 3 links
* 250 px link
* 500 px link
* optional 1200 px link
* "original" link (pointing to 500 px, 1200 px or high-res version)

so we can use these sizes for the preferences
* \>1200 px (if available)
* 1200 px (if available)
* 500 px <-- make that the default?
* 250 px

ronso0 commented:
One thing I noticed:
amongst the MusicBrainz results, the best tag match (for me) does not necessarily have the 'best'/correct cover art, so I'd actually appreciate to separate "Apply cover art" from "Apply tags".

Sorry if this has been discussed (and dismissed) earlier.
@fatihemreyildiz if we agree on that, probably this change can happen in a separate PR, maybe someone else can do this (me?) if you're short on time.

ronso0 commented:
> The cover size setting is hard to discover. Can we make them accessible from the tag fetcher dialog?

Yes, that would be nice, for example a combobox above the found cover but that would also make the TagFetcher dialog bigger.
Maybe a hint to choose the preferred size in the preferences would suffice, too. In  a collapsible 'info' section maybe?

ronso0 commented on src/musicbrainz/web/coverartarchivelinkstask.cpp:
```diff
@@ -0,0 +1,158 @@
+    emitSucceeded(allUrls);
+}
+
+void CoverArtArchiveLinksTask::emitSucceeded(
```
Why is this a separate method?
IIUC it's not called from anywhere else so it can be part of `onFinished` above, no?

fatihemreyildiz replied:
Sorry I've missed that review.

So this is separate method because AFAIK it is on that method it checks if the signal is connected to Tagfetcher just in case of any missing connections. Correct me please If I am wrong. On other tasks such as [acoustidlookuptask.cpp](https://github.com/mixxxdj/mixxx/blob/8affbbb2c5e429b9aa8e6ea7410b65c8a50e46b1/src/musicbrainz/web/acoustidlookuptask.cpp#L213) and [musicbrainzrecordingtask.cpp](https://github.com/mixxxdj/mixxx/blob/8affbbb2c5e429b9aa8e6ea7410b65c8a50e46b1/src/musicbrainz/web/musicbrainzrecordingstask.cpp#L199). I just went with the flow. 

There were two missing methods in image task. First missing method was for succeeded. It was okay I guess. But the failed signal was missing (IDK If we needed it at all, because all the failings are handled in Links task, if there were no links image task was not starting to work, if there are links there was no failing in image task), now it is added with the method 👍

fatihemreyildiz replied:
I couldn't fully understand, sorry. 

Should we remove the methods? If yes, In other tasks some of the methods are also used once. So should we also remove them too?

Thanks for sharing the YAGNI principle by the way.

fatihemreyildiz replied:
I will keep it like this for the symmetry 👍 

Thank you.

fatihemreyildiz committed:
dlgpreflibrary: Preference changed with pixels

fatihemreyildiz commented:
Thanks for all the testing and findings @ronso0 

> The json result from coverartarchive (for example https://ia800805.us.archive.org/17/items/mbid-cf93b2c3-9152-489d-b879-772ca46f5d81/index.json) always contains at least 3 links

AFAIK, all the responses at least had 2 links, 250px and 500px at all times. There is also image key in the response but as far as tested It was either 250px or 500px itself. 

So when users selected "Highest" resolution, they will get highest possible at all times. 
If they select "High", they will get either 1200 px (if available) or 500 px
For the medium and low, they will always receive their selection.

I will change the preferences with the pixels, so it will be easier to understand. 

> Sorry if this has been discussed (and dismissed) earlier.
@fatihemreyildiz if we agree on that, probably this change can happen in a separate PR, maybe someone else can do this (me?) if you're short on time.

We've discussed few different use-cases, and this was one of them. I also think that apply cover art would be better in use case too, but it will make the tag fetcher dialog a bit crowded. I think this can be done in separate PR later. What do you think? @daschuer 

>>The cover size setting is hard to discover. Can we make them accessible from the tag fetcher dialog?

>Yes, that would be nice, for example a combobox above the found cover but that would also make the TagFetcher dialog bigger.
Maybe a hint to choose the preferred size in the preferences would suffice, too. In a collapsible 'info' section maybe?

Combobox is great choice and it will improve the usability IMHO. We can still have preferences option (Preferred pixel), fetch the first cover art according to that, and the combobox can be changed for each tag, if the Combobox changed we can fetch the cover art again with the chosen item. On the other hand, this PR is big already and will it make this PR even bigger?

What do you think?

fatihemreyildiz commented:
Thank you for the findings @daschuer .

> * The cover size (pixel count) varies for differnt covers. I like to add covers roughly at the same size. Was this just a bad pick or is the cover size unpredictable? Can we give an estimation what the sizes mean in terms of pixel and bytes?
 
I couldn't fully understand for varying cover pixels. 

I think, I experienced once with one of the tags of `ABBA-Super Trouper` which was [this cover art.](https://ia800800.us.archive.org/17/items/mbid-2e428c4a-862b-4d39-bbb7-00a3b58814fc/mbid-2e428c4a-862b-4d39-bbb7-00a3b58814fc-21309259885.jpg). Which is in the album of "Gold: Greatest Hits". This is changing the orientation of the cover label on the tag fetcher. 

If It is the unpredictable cover art size, if this happens maybe centering it next to white columns would work. What do you think?

> * I think we need a different empty cover, for not available vs. not set.

That would be so nice, especially for the not available case, what comes to my mind at first is a title says "Not Found". What do you think @ronso0 ?

> * caching would be nice, for not fetching the same cover twice when scrolling though the results.

Do we need a caching class for that? If yes, should it be in a separate PR since this one is already big.

Is there a chance to keep this PR small and add the features in separate PR's?

fatihemreyildiz committed:
coverartfetcher: fix pre-commit.

ronso0 commented:
I think if there are no technical reasons holding this PR back we should merge it and address the remaining UX issues and the caching in separate PRs.
@fatihemreyildiz Any thoughts on https://github.com/mixxxdj/mixxx/pull/4851#pullrequestreview-1187338399

I've filed a few issues:
#11084 
#11085 
#11086

fatihemreyildiz committed:
covarartarchiveitasks: p prefix added for pointers

fatihemreyildiz committed:
coverartarchiveimagetask: missing emit methods add

fatihemreyildiz commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -270,6 +329,52 @@ void DlgTagFetcher::apply() {
+
+        QFileInfo trackFileInfo = QFileInfo(m_track->getLocation());
+
+        QString coverArtCopyFilePath =
+                trackFileInfo.absoluteFilePath().left(
+                        trackFileInfo.absoluteFilePath().lastIndexOf('.') + 1) +
+                "jpeg";
```
Thank you for the feedback @ronso0 

> why jpeg? can we be sure about that?

When I first implement this i used "jpg" because all of the responses were "jpg" on the links, that worked fine. After implementation of cover art copy worker, there was an error because the actual file extension was "jpeg" and on the name it was "jpg".  So I changed to jpeg.

> Idk if that's faster but it's more readable IMO.

That looks way better obviously. But with this change when we download the cover art there is one slash "/" is missing and that will download the cover art in wrong path. 

In my case the file paths are:
Old file path = "/home/username/Desktop/music/supertrouper.jpeg"
New file path =  "/home/username/Desktop/musicsupertrouper.jpeg"

When I implement it at first time, I wanted to use something easy to read in terms of file path, such as (canonicalpath() + '/' + trackBaseName + ".jpeg") but then I realized that might not work for windows since the file paths are separated with '\', I'm not sure about other operating systems.

So because of the various file path formats, I just deleted the extension of the track and added jpeg. I couldn't find a better way to compose the cover art file path, is there a better way to do this?

ronso0 replied:
Oh sorry, I didn't notice the missing dir separator.

Anyway, adding the suffix manually might not even be required. Please correct me if I'm wrong:
That you saw only `jpg` until now and no official info (I found) that musicbrainz only stores jpg... means we could get other formats and run into a format mismatch.
And then [`ImageFileData::saveFile`](https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/src/util/imagefiledata.cpp#L52) would throw an error but use the potentially wrong suffix anyway?
That's probably
> When I first implement this i used "jpg" because all of the responses were "jpg" on the links, that worked fine. After implementation of cover art copy worker, there was an error because the actual file extension was "jpeg" and on the name it was "jpg".

So, why specify the format at all when `ImageFileData` can figure out the correct extension from the fetched bytes array https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/src/util/imagefiledata.cpp#L12-L16 
Did I overlook something?

Btw [m_oldCoverArtFilePath](https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/src/library/export/coverartcopyworker.h#L27) should actually be m_**new**CoverArtFilePath, since that's the file that is written, no?

fatihemreyildiz replied:
Thanks for the review! There is no overlook, everything is fine. So no worries! 🐙 

I think everything is solved under this comment, except the [m_oldCoverArtFilePath](https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/src/library/export/coverartcopyworker.h#L27) . 

It is actually both new and old cover art file path. Because we have an existing cover art (old) and user is choosing new cover art (selected), then selected is replaced at where the old cover art is. So as I said it can be considered as both new and old. I'm not sure though. Should we rename it under this PR (Since I changed the related file on this PR already that will be ok IMHO), or this can be done in separate PR (A Pr with not big changes would make the PR history more crowded), or just not rename it at all. What do you think @ronso0 @daschuer ?

daschuer reviewed (changes requested):
I have found a bug during my last test. the rest is good.

fatihemreyildiz commented on src/musicbrainz/web/coverartarchivelinkstask.cpp:
```diff
@@ -0,0 +1,157 @@
+        emitFailed(response);
+        return;
+    }
+
```
I will simply add a link to a reference (A release with the most cover arts) instead of the whole JSON, that will look a bit better in shape, what do you think?

daschuer replied:
The link to a real Json is a good idea.
Please add also:
https://wiki.musicbrainz.org/Cover_Art_Archive/API

fatihemreyildiz commented on src/musicbrainz/web/coverartarchivelinkstask.cpp:
```diff
@@ -0,0 +1,157 @@
+        const auto thumbnails = imageObject.value(QLatin1String("thumbnails")).toObject();
+        DEBUG_ASSERT(!thumbnails.isEmpty());
+
+        const auto smallThumbnailUrl = thumbnails.value(QLatin1String("small")).toString();
```
This might fail and hit the assertions for the last tag in Abba - Super Trouper. 

The link of the release is: https://coverartarchive.org/release/5240094f-9d79-44fd-9985-77c7287bcc16

As far as I tested, I've only found this release as a different response than the others. Maybe there are others that might don't have small-large and have 250-500, I'm not sure.

Should we cover both "250" and "small" if small doesn't exists then try 250 or vice versa?

daschuer replied:
No, than don't change the code.
Just add a comment about our decision.

fatihemreyildiz replied:
Done, thank you!

fatihemreyildiz commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -399,4 +511,152 @@ void DlgTagFetcher::tagSelected() {
+        allUrls.size() < 3 ? getCoverArt(allUrls.last()) : getCoverArt(allUrls.at(2));
+        return;
+    } else if (fetcherQuality == DlgPrefLibrary::CoverArtFetcherQuality::Medium) {
+        allUrls.size() == 2 ? getCoverArt(allUrls.at(1)) : getCoverArt(allUrls.first());
```
This is way better block, thank you!

fatihemreyildiz committed:
dlgtagfetcher: fixed 500px cover art not fetching.

fatihemreyildiz committed:
dlgtagfetcher: compose fetched CA path improved

fatihemreyildiz committed:
coverartarchivelinkstask: example JSON schema add

ronso0 commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -337,10 +337,20 @@ void DlgTagFetcher::apply() {
-                trackFileInfo.absoluteFilePath().left(
-                        trackFileInfo.absoluteFilePath().lastIndexOf('.') + 1) +
-                "jpeg";
+        // This is the track file path without an extension.
+        // It is used to compose the Cover Art file path
+        // To prevent failure due to different directory slashes 
+        // QFileInfo::absoluteFilePath() is used.
+        // Everything after the latest '.' is deleted.
+        // eg trackFilePathWithoutSuffix: "/media/foo/Abba - Super Trouper."
+        QString trackFilePathWithoutSuffix = trackFileInfo.absoluteFilePath().left(
+                trackFileInfo.absoluteFilePath().lastIndexOf('.') + 1);
+
+        // ImageFileData::readFormatFrom returns the extension
+        // Extension of cover art will be the result returned from CoverArtArchive
+        // eg coverArtCopyFilePath: "/media/foo/Abba - Super Trouper.jpeg"
+        QString coverArtCopyFilePath = trackFilePathWithoutSuffix +
+                ImageFileData::readFormatFrom(m_fetchedCoverArtByteArrays);
```
Thanks.
Though, instead of creating and intermediate QString, let's compose the path in one step:
```suggestion
        // Compose the Cover Art file path. Use the correct file extension
        // by checking from the fetched image bytes (disregard extension
        // from file link).
        QString coverArtCopyFilePath = trackFileInfo.absoluteFilePath().left(
                trackFileInfo.absoluteFilePath().lastIndexOf('.') + 1) +
                ImageFileData::readFormatFrom(m_fetchedCoverArtByteArrays);
```

fatihemreyildiz replied:
Done! Thank you! 👍

ronso0 commented on src/library/dlgtagfetcher.cpp:
```diff
@@ -399,4 +514,145 @@ void DlgTagFetcher::tagSelected() {
     const int tagIndex = tags->currentItem()->data(0, Qt::UserRole).toInt();
     m_data.m_selectedTag = tagIndex;
+
+    if (!m_fetchedCoverArtByteArrays.isNull()) {
+        m_fetchedCoverArtByteArrays.clear();
+        m_pWFetchedCoverArtLabel->loadData(m_fetchedCoverArtByteArrays);
+        m_pWFetchedCoverArtLabel->setCoverArt(CoverInfo{},
+                QPixmap(CoverArtUtils::defaultCoverLocation()));
+    }
```
This seems to assume that if `m_fetchedCoverArtByteArrays` is null `m_pWFetchedCoverArtLabel` was already cleared (default cover image) and wants to avoid unnecessary clearing?
```suggestion
    m_fetchedCoverArtByteArrays.clear();
    m_pWFetchedCoverArtLabel->loadData(QByteArray());
    m_pWFetchedCoverArtLabel->setCoverArt(CoverInfo{},
    QPixmap(CoverArtUtils::defaultCoverLocation()));
```

fatihemreyildiz replied:
Yeah that was the purpose. But with the suggested changes it looks better

ronso0 reviewed (changes requested):
Some more findings while I was working on DlgCoverArtFullSize UX improvements.

fatihemreyildiz committed:
dlgtagfetcher: redundant disconnect deleted

fatihemreyildiz committed:
dlgtagfetcher: add fetched coverart always deleted

fatihemreyildiz committed:
dlgtagfetcher: p prefix added for pointers

ronso0 commented:
@daschuer Could you take another look?
Besides my recent reviews and some manual testing (went well, no issues), I didn't look into CoverArtArchiveLinkTask or CoverArtCopyWorker.

daschuer commented:
Sorry for the looong delay. I have now find time for a sorrowfully test and find no more issues. This is a good base for more. 
Thank you very much for that nice and useful feature.

daschuer commented:
@ronso0: Is your review also addressed?

ronso0 commented:
Yup, these issues are fixed, ready for merge in that regard.
(note I don't feel confident to click Approve or Merge since I touched only a small part of this branch)

ronso0 commented:
Yup, these issues are fixed, ready for merge in that regard.
(note I don't feel confident to click Approve or Merge since I touched only a small part of this branch)

daschuer commented:
Thank you for confirming. And thank you Emre for this nice and helpful feature.

fatihemreyildiz commented:
You are welcome Daniel! I'm glad that Mixxx has the cover art fetcher feature now. I hope everyone like this feature. 

I would like to also thank to everyone who reviewed the code and gave me feedback along the way. :octocat:

ronso0 commented:
:tada: 

Do we need some explanation in the manual, somewhere in [Library > Edit metadata of audio files](https://manual.mixxx.org/2.3/en/chapters/library.html#edit-metadata-of-audio-files)?

daschuer commented:
Yes
